### PR TITLE
Fix type rounding functions for MPFR

### LINKED ISSUE #33676: Inexact error when truncating from BigFloat to UInt8
`trunc(UInt8,255.1)` works fine but `trunc(UInt8,BigFloat(255.1))` throws an inexact error.

Tried it on Julia 1.2.0 for windows.

The relevant discussion on discord: https://discourse.julialang.org/t/strange-bug-when-truncating-bigfloat-to-uint8/30301/5

It seems the bug affects the truncation from `BigFloat` to any type of integer.

### PR DESCRIPTION
**Pull request by Keno**
Fixes #33676

### CODE REVIEW

**Comment by Keno** [comment]
bump

File: base/mpfr.jl [hunk 1]
```diff
@@ -263,67 +263,52 @@ BigFloat(x::AbstractString, r::RoundingMode; precision::Integer=DEFAULT_PRECISIO
...
-unsafe_cast(T, x::BigFloat, r::RoundingMode) = unsafe_cast(T, x, convert(MPFRRoundingMode, r))
+_unchecked_cast(T, x::BigFloat, r::RoundingMode) = _unchecked_cast(T, x, convert(MPFRRoundingMode, r))
 
-function unsafe_cast(::Type{Int64}, x::BigFloat, r::MPFRRoundingMode)
-    ccall((:__gmpfr_mpfr_get_sj,:libmpfr), Cintmax_t, (Ref{BigFloat}, MPFRRoundingMode), x, r)
-end
-function unsafe_cast(::Type{UInt64}, x::BigFloat, r::MPFRRoundingMode)
-    ccall((:__gmpfr_mpfr_get_uj,:libmpfr), Cuintmax_t, (Ref{BigFloat}, MPFRRoundingMode), x, r)
+function _unchecked_cast(::Type{Int64}, x::BigFloat, r::MPFRRoundingMode)
+    r = ccall((:__gmpfr_mpfr_get_sj,:libmpfr), Cintmax_t, (Ref{BigFloat}, MPFRRoundingMode), x, r)
```
Comment (simonbyrne) [inline base/mpfr.jl]: Why assign it to `r`?

File: base/mpfr.jl [hunk 2]
```diff
@@ -263,67 +263,52 @@ BigFloat(x::AbstractString, r::RoundingMode; precision::Integer=DEFAULT_PRECISIO
...
+    CT = T <: Signed ? Int64 : UInt64
+    typemax(T) < typemax(CT) ? _unchecked_cast(CT, x, r) : _unchecked_cast(BigInt, x, r)
 end
 
-function round(::Type{T}, x::BigFloat) where T<:Union{Signed,Unsigned}
-    (typemin(T) <= x <= typemax(T)) || throw(InexactError(:round, T, x))
-    unsafe_cast(T, x, ROUNDING_MODE[])
+function round(::Type{T}, x::BigFloat, r::Union{RoundingMode, MPFRRoundingMode}) where T<:Union{Signed, Unsigned}
+    clear_flags()
+    r = _unchecked_cast(T, x, r)
```
Comment (simonbyrne) [inline base/mpfr.jl]: maybe assign to a symbol other than `r`?

File: base/floatfuncs.jl [hunk 3]
```diff
@@ -120,8 +120,10 @@ To extend `round` to new numeric types, it is typically sufficient to define `Ba
 """
 round(T::Type, x)
 
-round(::Type{T}, x::AbstractFloat, r::RoundingMode{:ToZero}) where {T<:Integer} = trunc(T, x)
-round(::Type{T}, x::AbstractFloat, r::RoundingMode) where {T<:Integer} = trunc(T, round(x,r))
+function round(::Type{T}, x::AbstractFloat, r::RoundingMode) where {T<:Integer}
+    r != RoundToZero && (x = round(x,r))
+    trunc(T, x)
+end
```
[the remark this answers is not in the document]
Comment (Keno) [reply base/floatfuncs.jl]: IIRC it caused a bunch of unnecessary ambiguities.

File: base/mpfr.jl (hunk 2 above)
Comment (Keno) [reply base/mpfr.jl]: good catch

File: base/mpfr.jl [hunk 4]
```diff
@@ -263,67 +263,52 @@ BigFloat(x::AbstractString, r::RoundingMode; precision::Integer=DEFAULT_PRECISIO
...
 
 ## BigFloat -> Integer
-"""
-    MPFR.unsafe_cast(T, x::BigFloat, r::RoundingMode)
-
-Convert `x` to integer type `T`, rounding the direction of `r`. If the value is not
-representable by T, an arbitrary value will be returned.
-"""
-unsafe_cast(T, x::BigFloat, r::RoundingMode) = unsafe_cast(T, x, convert(MPFRRoundingMode, r))
+_unchecked_cast(T, x::BigFloat, r::RoundingMode) = _unchecked_cast(T, x, convert(MPFRRoundingMode, r))
```
[the remark this answers is not in the document]
Comment (Keno) [reply base/mpfr.jl]: My intention is for people not to use this function directly, which is why I removed the docs and added the leading `_`. The mpfr functions it calls are well documented, so any local users touching this code should just look at those and trust those docs over anything we write here. If you feel strongly about it though, I can put back the docs.

**Comment by StefanKarpinski** [comment]
Bump?

**Comment by JeffreySarnoff** [comment]
rebump

File: base/mpfr.jl (hunk 1 above)
Comment (simonbyrne) [reply base/mpfr.jl]:
```suggestion
    ccall((:__gmpfr_mpfr_get_sj,:libmpfr), Cintmax_t, (Ref{BigFloat}, MPFRRoundingMode), x, r)
```

**Comment by simonbyrne** [comment]
LGTM with the above two minor changes.

**Pull request merged**